### PR TITLE
Fix gen_domain tool to work in the absence of masks

### PR DESCRIPTION
**Pull request by brhillman**
Remove dependence on "mask" variables in the mapping files that are input to the `gen_domain` 
tool.

The `gen_domain` tool is used to construct "domain files", which indicate the fractional coverage of component grids. This tool previously expected "mask" variables for both the target and source grids, which indicated whether or not component values would be valid at each point in the grid. This concept applied to FV grids, where a global mesh was established but various components may or may not be active at each point in that mesh. For example, the ocean would not be active over land points, so the mask should be zero for those points. This concept does not necessarily apply to newer unstructured grids, which can be designed to simply not include points where the component would not be active. Thus, a mask variable is unnecessary for these grids.

Add options to set the ocean mask to 1 if its not present.  Program will still die if mask is not present and --set-omask is not given.

Also make the fminval and fmaxval optional input paramaters.

### CODE REVIEW

**Comment by paullric** [comment]
Seems this pull will still output an error to stdout if “mask_a” and “mask_b” are not present, but won’t cause the program to abort.  Maybe it should be a warning instead?

Also with the latest fix to GenerateTransposeMap the “domain_a” and “domain_b” attributes should be retained, so perhaps these should be required?

Otherwise looks good.

~ Paul

**Comment by brhillman** [comment]
The error is output if "domain_a" and "domain_b" are not present, but it passes over missing "mask_a" and "mask_b" silently (`if (var_exists(fid, 'mask'//trim(suffix)))` ensures this).

**Comment by brhillman** [comment]
Thanks @czender. @rljacob anything else I need to do to get this merged? Thanks.

**Comment by rljacob** [comment]
What kind of testing have you done?

**Comment by brhillman** [comment]
@rljacob that's a good question. I did a basic sanity check to make sure that my changes here did not affect domain files created when "mask" variables are present in the mapping files...i.e., when using maps generated using ESMF. For this, I just used a small ne4 atmosphere to oQU240 ocean example, generated domain files from the same mapping file using the code from master and then separately from this branch, and then compared the values in the result netcdf files and found them to be identical. So I'm reasonably convinced I have not broken previous functionality, but we do not have rigorous tests setup for any of these tools as far as I know.

To try to understand whether or not the behavior I've implemented is reasonable for domain files generated from Tempest mapping files, I did some minimal analysis looking at the variables in the domain files generated using ESMF vs Tempest masks, and I'm a little confused by the results and really I probably need some more expert eyes and opinions to evaluate this, but I am not sure who to ask within the project. There are three relevant fields in the domain files: area, frac, and mask. Comparing area and frac between domains files generated from ESMF vs Tempest masks appear to be *consistent* at least, but numerically will be slightly different because of the difference in mapping files. However, the mask variable looks quite different...and I'm not sure how this is turning out the way it is for the tempest case. See attached figures for comparison for `domain.ocn.ne4np4_oQU240` files:

domain_area
domain_frac
domain_mask

**Comment by paullric** [comment]
Based on a cursory examination, it seems that this issue is related to the fact that ESMF uses the dual grid, which is a finite volume mesh built around each GLL node, whereas Tempest uses the full spectral element.  Consider this in the context of a “domain of influence” associated with each mesh cell:  The finite volume grid cell would be pretty small and hence mapping would be much more local to that particular degree of freedom; on the other hand, for the Tempest maps a single GLL node can influence the target data from as far away as a single spectral element (a radius 3-6x larger than the radius of the finite volume).

Based on this logic I would hypothesize these results would converge as ne -> large.

~ Paul

**Comment by rljacob** [comment]
@iulian787 can you look at this?

**Comment by brhillman** [comment]
@paullric okay so there are more "unmasked" points (mask = 1) in the tempest case because the "domain of influence" is larger for each GLL node...I think I get what you're saying, and why there are much more mask=1 points in the tempest example here. This is also consistent with what I see in the `domain.lnd` files (see attached figure showing just the "mask" variable): it takes a larger area to complete mask out the domain in the tempest case. The question I guess is, is this going to affect the simulation? How is "mask" from the domain files used at runtime? I will do an ne30 example for comparison as well.

domain_mask_lnd

**Comment by rljacob** [comment]
So the "domain_mask" variable is only seeing the most land-locked land?

**Comment by brhillman** [comment]
@rljacob that's my interpretation, or the most ocean-locked ocean in the inverse file.

**Comment by rljacob** [comment]
How is "domain_mask" used vs. "fraction of grid cell that is active".

**Comment by bishtgautam** [comment]
As I understand, values in `mask` correspond to the land/ocean mask and it is used by the land model to determine the number of active grid cells. While the values in `frac` corresponds to the fraction of the active grid cell that is land.

**Comment by rljacob** [comment]
Ok.  "mask" is also an integer array in the few domain files I looked at so it should be 1 wherever frac >0 and 0 otherwise.

**Comment by thorntonpe** [comment]
In the subroutine initialize1(), within clm_initializeMod.F90, there is a call to surfrd_get_globmask() (in surfrdMod.F90). That call is where the land model gets its information about the number of land gridcells. The code looks for a variable named "mask" or "LANDMASK" on the fatmlndfrc file (which is CLM's internal name for the domain file). At this stage, the land fraction information is not used. All of the CLM grid and subgrid construction is based on the number of cells with the value "1" in this mask.

**Comment by brhillman** [comment]
So it seems that "mask" is what is important here, and that is what seems to not be calculated quite right for tempest mapping files right now...hmm. I need to take a more careful look at this to understand how mask is being (or should be) determined when not provided by the mapping files.

**Comment by rljacob** [comment]
@jonbob how did you generate domain files with MPAS grids that don't have any land in them?

**Comment by paullric** [comment]
One more follow-up that came to mind:  You can reduce the radius of influence for every GLL node in Tempest by using a monotonic map (command line option --mono).  I suspect with this option your results will look a lot more reasonable.

~ Paul

**Comment by rljacob** [comment]
@jonbob this PR has some questions that need your answers.

**Comment by jonbob** [comment]
@rljacob - all of the MPAS ocn/ice grids have implied land in them, don't they? At least in this approach where we assume land is wherever ocn/ice is not. So I just used this tool with the ESMF aave ocn=>atm map

**Comment by mt5555** [comment]
I believe this is the expected behavior:  I believe that the component models will be using the factional mask and not the domain mask.  In addition, as Paul says, this will improve dramatically under mesh refinement.  We evaluated this mask with TR maps at NE30 a year or so ago, and it looked good (with the transpose map)

**Comment by czender** [comment]
Note the utility of propagating an accurate `history` attribute in mapfiles.

**Comment by brhillman** [comment]
These certainly look better for ne30, but the "domain mask" still looks a little goofy to me:
screen shot 2019-01-15 at 1 18 54 pm

If the components are using the fractional mask instead, then this seems perfectly fine, but if they are using the domain mask then this seems a little sketchy maybe?

**Comment by brhillman** [comment]
Note these are still with the oQU240 ocean grid, I will repeat with a different ocean grid.

**Comment by brhillman** [comment]
This definitely doesn't work with a gx1v6 grid:
screen shot 2019-01-15 at 1 47 42 pm
So something weird is going on.

**Comment by mt5555** [comment]
could this be a precision issue?  domain mask should be points where domain fraction > 1-tol, and it looks like tol should be increased.

**Comment by brhillman** [comment]
@mt5555 I don't think so, because in that last figure with gx1v6, the domain fraction goofy too. There is a `fminval` in the code against which domain fraction is compared to determine whether or not to reset domain mask to 0. This was hard-coded as `fminval = 0.001`. I tried setting this to `fminval = 0` and it affects the domain mask from the ESMF mapping file but does not seem to affect the issue here with the domain files created from tempest maps:
screen shot 2019-01-15 at 2 34 28 pm

**Comment by rljacob** [comment]
@brhillman how are you making those plots?  Does it involve any additional interpolation to lat lon grids?

**Comment by brhillman** [comment]
@rljacob I plot using `tripcolor` in matplotlib, so there is a level of implicit interpolation. I am plotting the mask and area in the domain files directly, with the lat/lon of the grid cell centers to define the coordinates. I.e., `area` has dimensions `nj, ni`, and `xc(nj, ni), yc(nj, ni)` define the lat/lons. Code to do this in python/matplotlib:
```
from xarray import open_dataset
from matplotlib import pyplot
from cartopy import crs

# Define a function to plot a single map
def plot_map(x, y, data, **kwargs):
    ax = pyplot.gca()
    pl = ax.tripcolor(x, y, data, **kwargs)
    return pl


# Define a function to plot maps from multiple datasets
def compare_maps(datasets, labels, field_name, **kwargs):
    
    # Setup figure
    figure, axes = pyplot.subplots(1, len(datasets))
    
    # Find mins and maxes to setup a common plot range
    vmax = max([ds[field_name].max().values for ds in datasets])
    vmin = min([ds[field_name].min().values for ds in datasets])
    
    # We get weird results if vmin == vmax, so make a real range here
    if vmin == vmax:
        vmin = 0.9 * vmin
        vmax = 1.1 * vmax
    
    # Make plots
    for icase, (ds, case) in enumerate(zip(datasets, labels)):
        
        # Select data
        x = ds['xc'].squeeze()
        y = ds['yc'].squeeze()
        z = ds[field_name].squeeze()
        
        # Select axes and make plot
        ax = figure.add_axes(axes.ravel()[icase])
        pl = plot_map(x, y, z, vmin=vmin, vmax=vmax, **kwargs)
            
        # Label this plot axes
        ax.set_title('%s'%(case))
        
    # Add a colorbar
    cb = pyplot.colorbar(
        pl, ax=axes.ravel().tolist(), orientation='horizontal',
        label='%s'%(z.long_name),
    )
    
    # Return figure handle
    return figure

cases = ('esmf ocean-land (ne30)', 'tempest ocean-land (ne30)')
for field_name in ('area', 'mask', 'frac'):
    datasets = [open_dataset(domain_files[case]) for case in cases]
    figure = compare_maps(datasets, cases, field_name)
```

**Comment by mt5555** [comment]
actually, the gx1v6 results are bad, even for the fractional version - that looks like some kind of bug?

**Comment by brhillman** [comment]
Yeah @mt5555 I agree. I'm not sure what's going on. I'm looking through the code but nothing is jumping out. Because the fractional version looks bad, I don't think it is something to do with `fminval`.

**Comment by brhillman** [comment]
@mt5555 okay, I think I understand why this fails for the gx1v6 grid. The problem is that the gx1v6 grid is a regular global grid, so the trick in the domain code to set the mask to 1 at all points in the domain in the absence of a valid mask variable in the mapping file doesn't work. So I think the problem is that TempestRemap does not propagate the mask variable from the source grid (the gx1v6 ocean grid in this case). So while this does not matter for MPAS grids that are *probably* only defined where the component is active (i.e., ocean points only where there is ocean), it does seem to matter for other grids. This depends of course on how the component models are using the domain file. If they are using it to determine actual land/ocean fractions, then this might be problematic. The fix is probably to have TempestRemap propagate the mask variable from SCRIP-style grids that it reads, if this variable is present. For example, for the gx1v6 grid, this is named `grid_imask` in the SCRIP file, and ESMF propagates this to the mapping file (i.e., `map_gx1v6_to_ne30np4_aave.190115.nc`) as `mask_a`. Pinging @paullric because I think the fix is to have TempestRemap propagate these masks to the mapping files, at least so that our approach here is general and backwards compatible. One oddity here, is that I still don't understand why the domain code generates any non-zero land fraction at all with maps from Tempest in this case, because it seems to me that the ocean domain should be "1" everywhere with this approach.

**Comment by mt5555** [comment]
great work tracking this down.  you might also be able to 'ncks' mask_a from the SCRIP file into the mapping file after running TR?

**Comment by brhillman** [comment]
Note @paullric recent fixes in Tempest to propagate masks from grid files fixes this for current configurations, where at least the non-SE grids will have a mask present. However, I think we still need to allow gen_domain to work in the absence of masks, because in the case that neither source nor destination grid files contain masks, new masks will not be created and so inferring masks as we do in the changes in this PR will probably be necessary.

**Comment by brhillman** [comment]
Another possible issue here when using this with Tempest maps, is that it appears that remapping the ocean mask to the atmos grid might result in some non-negligible errors in the computed land and ocean fraction that might be cause for concern. Output from the code shows the original min and max of the computed land fraction, which is clipped by the code if below or above some threshold as @mt5555 mentioned. But examining the values here from domains created with ESMF vs Tempest mapping files suggests a much larger difference in the remapped masks; for the oQU240 to ne4 case, using the ESMF conservative mapping, the original min of the land fraction (computed as `1 - ofrac`) is -4.4e-15, but with the Tempest monotr mapping the min of the land fraction (before clipping) is -2.8e-4. Is this cause for concern? Note that in the original `gen_domain` code, there is a now-unused variable `eps` which supposedly at one point set the threshold for the tolerance for this error (and is written to the output netcdf files, but is not actually used anywhere), with a default value of 1e-12. But, the error in domains with maps made from ESMF for the gx1v6 to ne30 case (which we currently use for F-compsets it looks like) exceeds this (1e-11 it looks like), so I'm not sure how much to worry about that. Pinging @paullric and @mt5555 for their mapping expertise.

File: cime/tools/mapping/gen_domain_files/src/gen_domain.F90 [hunk 1]
```diff
@@ -335,9 +346,14 @@ subroutine gen_domain(fmap, fn1_out_ocn, fn2_out_lnd, fn2_out_ocn, set_fv_pole_y
...
-         call check_ret(nf_inq_varid(fid,'mask'//trim(suffix), vid ))
-         call check_ret(nf_get_var_int   (fid,vid,omask ))
+         ! Determine ocn mask on ocn grid; if mask is not present, we assume
+         ! that the grid covers the entire domain, so mask = 1 for all points.
+         if (var_exists(fid, 'mask'//trim(suffix))) then
+            call check_ret(nf_inq_varid(fid,'mask'//trim(suffix), vid ))
+            call check_ret(nf_get_var_int   (fid,vid,omask ))
+         else
+            omask(:) = 1
+         end if
```
Comment (rljacob) [inline cime/tools/mapping/gen_domain_files/src/gen_domain.F90]: Bill, after talking about this with CESM, can you implement this differently?   Add another optional argument called something like "--noomask".  If its present AND the mask variable does not exist, then set omask to 1.   Error if --nomask is set AND the variable DOES exist.  Or if the variable is doesn't exist and --nomask is not set.  Similar for mask_a below.

Comment (rljacob) [reply cime/tools/mapping/gen_domain_files/src/gen_domain.F90]: maybe "--setomask" and "--setamask" would be better names.

**Comment by rljacob** [comment]
@brhillman can you make the changes requested above?  I'd like to finish this PR.

**Comment by brhillman** [comment]
@rljacob sure, I can get to this later today, but I think with @paullric changes to Tempest this PR is probably no longer required. (By the way, it's Ben not Bill)

File: cime/tools/mapping/gen_domain_files/src/gen_domain.F90 (hunk 1 above)
Comment (brhillman) [reply cime/tools/mapping/gen_domain_files/src/gen_domain.F90]: Rob, what about “—no-omask” and “—no-amask”, and if unset and mask does not exist, throw an error, but IF SET, then read masks if they exist and set equal to 1 if they do not? Is there a reason to throw an error if the flag is set and the mask actually already exists? I get wanting to throw an error by default when the flag is unset and the masks don’t exist, that totally makes sense.

Comment (rljacob) [reply cime/tools/mapping/gen_domain_files/src/gen_domain.F90]: The if test is "do something if mask is not present".  So we instead want an option which specifies that action.  The action to take is "set the mask".  So the option should be "set-omask"  "set-amask".

Yeah I guess you don't have to flag an error if the flag is set and the mask is present.

Comment (brhillman) [reply cime/tools/mapping/gen_domain_files/src/gen_domain.F90]: I have this implemented, just need to test. But...should we just omit this addition and either close this PR, or just include the fminval and fmaxval additions and raise an error if mask doesn’t exist? Now that Paul made the changes in tempest to propagate mask and domain, we shouldn’t need to worry about mask not being present.

**Comment by brhillman** [comment]
@rljacob I have your requested changes implemented and just need to test on my local branch, but should we just close this PR since Tempest now propagates maps and attributes properly? Or at least require the masks be present, but accept the addition added here to allow setting fminval and fmaxval?

**Comment by rljacob** [comment]
That is your call.   If you just want to make the fminval, fmaxval changes, probably best to close this a make a new branch.

**Comment by brhillman** [comment]
@rljacob changes made and commit pushed. I tested that default behavior gives the same answers (against master) for oRRS30to10_to_ne30np4, gx1v6_to_ne30np4, and oQU240_to_ne4np4 aave mapping files. I then tested the optional behavior (enabled by passing --set-omask) by creating new mapping files for oRRS30to10_to_ne30np4 via `ncks -x -v mask_a` to drop the ocean mask and then recreating the domain files via `gen_domain --set-omask`. This produced identical answers. I also confirmed that omitting`--set-omask` when `mask_a` is not present raises the error and the program exits without generating the domain files. Note that `mask_b` is never read in this code (I also confirmed this by creating another mapping file without `mask_b`), so `--set-omask` seemed to be an appropriate name for this option.

**Comment by rljacob** [comment]
@paullric and @jonbob do you approve this PR?

**Comment by rljacob** [comment]
@paullric your review is requested on this PR.

**Comment by rljacob** [comment]
Not waiting for last review.  Merged to next.

**Comment by rljacob** [comment]
no testing for this so merged to master.

**Pull request merged**